### PR TITLE
Interpolation Performance part 1

### PR DESCRIPTION
**Pull request by Kenneth-T-Moore**
### Summary

1. Improved the performance of the "cubic" method by replacing instances of vector division with multiplication.
2. Added two new interpolation methods "akma1d" and "trilinear" that are restricted to fixed table dimensions but are much faster.

### Related Issues

- Resolves #

### Backwards incompatibilities

None

### New Dependencies

None

### CODE REVIEW

File: openmdao/components/interp_util/interp_algorithm.py [hunk 1]
```diff
@@ -239,6 +258,282 @@ def interpolate(self, x, idx, slice_idx):
...
+        ndarray
+            Interpolated values.
+        ndarray
+            Derivative of interpolated values with respect to independents.
+        ndarray
+            Derivative of interpolated values with respect to values.
+        ndarray
+            Derivative of interpolated values with respect to grid.
+        """
+        pass
```
Comment (naylor-b) [inline openmdao/components/interp_util/interp_algorithm.py:534]: This should probably raise a NotImplementedError

**Review by naylor-b — CHANGES_REQUESTED** [review]

File: openmdao/components/interp_util/interp_akima.py [hunk 2]
```diff
@@ -1292,3 +1293,222 @@ def interpolate(self, x):
...
+            dtype = self.values.dtype
+        else:
+            dtype = x.dtype
+
+        # Check for extrapolation conditions. if off upper end of table (idx = ient-1)
+        # reset idx to interval lower bracket (ient-2). if off lower end of table
+        # (idx = 0) interval lower bracket already set but need to check if independent
+        # variable is < first value in independent array
+        ngrid = len(grid)
+        if idx == ngrid - 1:
```
[the remark this answers is not in the document]
Comment (Kenneth-T-Moore) [reply openmdao/components/interp_util/interp_akima.py:1384]: I did a little testing, and `idx = ngrid - 2` is actualy a little faster than `idx -= 1`.  
Made the other change though.

File: openmdao/components/interp_util/interp_akima.py [hunk 3]
```diff
@@ -1292,3 +1293,222 @@ def interpolate(self, x):
...
+        # m4 is the slope of interval (xi+1, xi+2)
+        # m5 is the slope of interval (xi+2, xi+3)
+        #
+        # The values of m1, m2, m4 and m5 may be calculated from other slope values
+        # depending on the value of idx
+
+        val3 = values[idx]
+        val4 = values[idx + 1]
+        h = 1.0 / (grid[idx + 1] - grid[idx])
+        m3 = np.zeros(1, dtype=dtype)
```
[the remark this answers is not in the document]
Comment (Kenneth-T-Moore) [reply openmdao/components/interp_util/interp_akima.py:1448]: Replaced `zeros` with `empty`. 
That quantity definitely won't be a scalar once we add vectorization (unless we loop on the coeff computations), so I'll make a note on looking into that.

File: openmdao/components/interp_util/interp_akima.py [hunk 4]
```diff
@@ -1292,3 +1293,222 @@ def interpolate(self, x):
...
+        m3[:] = (val4 - val3) * h
+
+        if idx >= 1:
+            val2 = values[idx - 1]
+            m2 = np.zeros(1, dtype=dtype)
```
[the remark this answers is not in the document]
Comment (Kenneth-T-Moore) [reply openmdao/components/interp_util/interp_akima.py:1453]: replaced all 5

File: openmdao/components/interp_util/interp_trilinear.py [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+        ----------
+        x : float
+            Value of new independent to interpolate.
+
+        Returns
+        -------
+        bool
+            Returns True if this table can be run vectorized.
+        """
+        if x.shape[0] > 1:
```
[the remark this answers is not in the document]
Comment (Kenneth-T-Moore) [reply openmdao/components/interp_util/interp_trilinear.py:60]: replaced

File: openmdao/components/interp_util/interp_algorithm.py (hunk 1 above)
Comment (Kenneth-T-Moore) [reply openmdao/components/interp_util/interp_algorithm.py:534]: Replaced

**Review by naylor-b — APPROVED** [review]

**Review by swryan — APPROVED** [review]

**Pull request merged**